src/unistd.rs

```diff
@@ -941,6 +941,158 @@ pub fn setgid(gid: Gid) -> Result<()> {
     Errno::result(res).map(drop)
 }
 
+/// Get the list of supplementary group IDs of the calling process.
+///
+/// *Note:* On macOS, `getgroups()` behavior differs somewhat from other Unix
+/// platforms. It returns the current group access list for the user associated
+/// with the effective user id of the process; the group access list may change
+/// over the lifetime of the process, and it is not affected by calls to
+/// `setgroups()`.
+///
+/// [Further reading](http://pubs.opengroup.org/onlinepubs/009695399/functions/getgroups.html)
+pub fn getgroups() -> Result<Vec<Gid>> {
+    // First get the number of groups so we can size our Vec
+    use std::ptr;
+    let ret = unsafe { libc::getgroups(0, ptr::null_mut()) };
+    let mut size = try!(Errno::result(ret));
+
+    // Now actually get the groups
+    let mut groups = Vec::<Gid>::with_capacity(size as usize);
+    loop {
+        // We can coerce a pointer to some `Gid`s as a pointer to some `gid_t`s
+        // as they have the same representation in memory.
```

Comment:
This is actually not technically true. newtype's are not guaranteed to have the same memory layout as their containing structs. `repr(transparent)` has been merged as [an RFC](https://github.com/rust-lang/rfcs/blob/master/text/1758-repr-transparent.md) to rectify this. We've been making this assumption within nix because I think for all platforms we actually target it's true, but we need to be careful with this.